backend/data/src/main/kotlin/io/tolgee/service/UserAccountService.kt

```diff
@@ -202,35 +244,51 @@ class UserAccountService(
     }
   }
 
-  fun encodePassword(rawPassword: String): String {
-    val bCryptPasswordEncoder = BCryptPasswordEncoder()
-    return bCryptPasswordEncoder.encode(rawPassword)
-  }
-
   @Transactional
   @CacheEvict(cacheNames = [Caches.USER_ACCOUNTS], key = "#result.id")
   fun update(userAccount: UserAccount, dto: UserUpdateRequestDto): UserAccount {
+    // Current password required to change email or password
+    if (dto.email != userAccount.username) {
+      if (userAccount.accountType == UserAccount.AccountType.LDAP) {
+        throw BadRequestException(Message.OPERATION_UNAVAILABLE_FOR_ACCOUNT_TYPE)
+      }
+
+      if (dto.currentPassword?.isNotEmpty() != true) throw BadRequestException(Message.CURRENT_PASSWORD_REQUIRED)
+      val matches = passwordEncoder.matches(dto.currentPassword, userAccount.password)
+      if (!matches) throw PermissionException()
+    }
+
     val old = UserAccountDto.fromEntity(userAccount)
     updateUserEmail(userAccount, dto)
-    updatePassword(dto, userAccount)
     userAccount.name = dto.name
+
     publishUserInfoUpdatedEvent(old, userAccount)
     return userAccountRepository.save(userAccount)
   }
 
-  private fun updatePassword(dto: UserUpdateRequestDto, userAccount: UserAccount) {
-    dto.password?.let {
-      if (!it.isEmpty()) {
-        userAccount.password = encodePassword(it)
-      }
+  fun updatePassword(userAccount: UserAccount, dto: UserUpdatePasswordRequestDto): UserAccount {
+    if (userAccount.accountType == UserAccount.AccountType.LDAP) {
+      throw BadRequestException(Message.OPERATION_UNAVAILABLE_FOR_ACCOUNT_TYPE)
     }
+
+    val matches = passwordEncoder.matches(dto.currentPassword, userAccount.password)
+    if (!matches) throw PermissionException()
+
+    userAccount.tokensValidNotBefore = DateUtils.truncate(Date(), Calendar.SECOND)
+    userAccount.password = passwordEncoder.encode(dto.password)
+    return userAccountRepository.save(userAccount)
   }
 
   private fun updateUserEmail(
     userAccount: UserAccount,
     dto: UserUpdateRequestDto
   ) {
     if (userAccount.username != dto.email) {
+      if (!Regex("[^@]+@[^@]+").matches(dto.email)) {
```

Comment:
Maybe we can only make email nullable in the request and when not provided, we can leave it, so we can keep the email check there. This is not a good way how to check e-mail anyway...

Reply:
I used the bare minimum regex to check emails, the common rationale around this being that it's useless to check for a stricy regex which is heavier to process and will still let invalid emails through (or for some regexes, reject actually valid emails because they make wrong assumptions). There are a lot of articles about it online

I can use a stricter regex, or look into manually calling hibernate validator if that's too much of a concern

Reply:
Call the validator. This can cause another exceptions thrown while sending e-mails to invalid e-mails.